### PR TITLE
Move GradientController to top-level components

### PR DESCRIPTION
**Pull request by hasit**
### Description of change

- Remove controller interface and related files
- Migrate GradientController to be a top-level component

screencapture-localhost-3000-d-f5fafd6dbdc3fef42960b4d75615afd5463b2564-aperture-dashboard-service1-demo-app-2023-10-23-10_30_53

##### Checklist

- [x] Tested in playground or other setup
- [x] Screenshot (Grafana) from playground added to PR for 15+ minute run
- [x] Documentation is changed or added
- [x] Tests and/or benchmarks are included
- [ ] Breaking changes

### Summary by CodeRabbit

- Refactor: Simplified and reorganized the codebase for better maintainability. This includes changes in the import statements and function calls in the `component-factory.go` and `component-factory_test.go` files.
- New Feature: Introduced a new `GradientController` component. This component enhances the software's ability to handle dynamic configuration updates and compute outputs based on input readings.
- Style: Made cosmetic changes to the `PIDController` struct and its associated functions for improved readability and consistency.
- Chore: Removed unused import statements in the `dashboard_group.libsonnet` and `main.libsonnet` files, indicating a cleaner and more efficient codebase.

### CODE REVIEW

File: pkg/policies/controlplane/components/gradient-controller.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+// IsActuator implements runtime.Component.
+func (g *GradientController) IsActuator() bool {
+	return false
+}
+
+// Execute implements runtime.Component.Execute.
+func (g *GradientController) Execute(inPortReadings runtime.PortToReading, circuitAPI runtime.CircuitAPI) (outPortReadings runtime.PortToReading, err error) {
+	signal := inPortReadings.ReadSingleReadingPort("signal")
+	setpoint := inPortReadings.ReadSingleReadingPort("setpoint")
+	optimize := inPortReadings.ReadSingleReadingPort("optimize")
```
Comment (tanveergill) [inline pkg/policies/controlplane/components/gradient-controller.go:80]: Don't have `optimize` port on `GradientController`

File: pkg/policies/controlplane/components/gradient-controller.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+		}
+	}
+
+	// Set output to control variable in-case of Manual mode
+	if g.manualMode {
+		output = controlVariable
+	}
+
+	// Save readings for the next tick so that Controller may access them via ControllerStateReadAPI
+	g.lastOutput = output
```
Comment (tanveergill) [inline pkg/policies/controlplane/components/gradient-controller.go:132]: No need to remember the last reading

**Review by tanveergill — APPROVED** [review]
Can merge after addressing comments

File: pkg/policies/controlplane/components/gradient-controller.go (hunk 1 above)
Comment (hasit) [reply pkg/policies/controlplane/components/gradient-controller.go]: [63c267f](https://github.com/fluxninja/aperture/pull/2838/commits/63c267ff138b9e85f4f18813a91aa8bf78e58e8e) addresses this.

File: pkg/policies/controlplane/components/gradient-controller.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+		}
+	} else {
+		output = runtime.InvalidReading()
+	}
+
+	// Optimize
+	if output.Valid() && optimize.Valid() {
+		targetOutput := runtime.NewReading(output.Value() + optimize.Value())
+		output = targetOutput
+	}
```
[the remark this answers is not in the document]
Comment (hasit) [reply pkg/policies/controlplane/components/gradient-controller.go]: [63c267f](https://github.com/fluxninja/aperture/pull/2838/commits/63c267ff138b9e85f4f18813a91aa8bf78e58e8e) addresses this.

File: pkg/policies/controlplane/components/gradient-controller.go (hunk 2 above)
Comment (hasit) [reply pkg/policies/controlplane/components/gradient-controller.go]: [63c267f](https://github.com/fluxninja/aperture/pull/2838/commits/63c267ff138b9e85f4f18813a91aa8bf78e58e8e) addresses this.

File: pkg/policies/controlplane/components/gradient-controller.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+	// Set output to control variable in-case of Manual mode
+	if g.manualMode {
+		output = controlVariable
+	}
+
+	return runtime.PortToReading{
+		"output": []runtime.Reading{output},
+	}, nil
+}
```
[the remark this answers is not in the document]
Comment (hasit) [reply pkg/policies/controlplane/components/gradient-controller.go]: [e8a56df](https://github.com/fluxninja/aperture/pull/2838/commits/e8a56dfe2c7b0453e6480cc8707c87d941e5962c) addresses this.

**Pull request merged**